File: longest_substr/longest_substr.cc
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int sidx = 0, ans = 0;
        unordered_map<char, int> m;
        for (int i = 0; i < s.length(); ++i) {
            if (m.find(s[i]) != m.end() && m[s[i]] >= sidx) {
                int l = i - sidx;
                ans = max(ans, l);
                sidx = m[s[i]] + 1;
            }
            m[s[i]] = i;
            if (i == s.length() - 1){
                int l = i - sidx + 1;
                ans = max(ans, l);
            }
        }
        return ans;
    }
};

File: longest_substr/longest_substr.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start_i = ans = 0
        d = {}
        
        for i in range(len(s)):
            if s[i] in d and d[s[i]] >= start_i:
                l = i - start_i
                ans = max(ans, l)
                start_i = d[s[i]] + 1
                d[s[i]] = i
            
            d[s[i]] = i
            
            if i+1 == len(s):
                l = i - start_i + 1
                ans = max(ans, l)
                
        return ans
